src/flint/types/fmpq_mpoly.pyx

```diff
@@ -336,205 +352,77 @@ cdef class fmpq_mpoly(flint_mpoly):
-                return res
-        return NotImplemented
-
-    def __radd__(self, other):
-        cdef fmpq_mpoly res
-        other = any_as_fmpq(other)
-        if other is not NotImplemented:
-            res = create_fmpq_mpoly(self.ctx)
-            fmpq_mpoly_add_fmpq(res.val, (<fmpq_mpoly>self).val, (<fmpq>other).val, res.ctx.val)
-            return res
-        return NotImplemented
 
-    def __sub__(self, other):
-        cdef fmpq_mpoly res
-        if typecheck(other, fmpq_mpoly):
-            if (<fmpq_mpoly>self).ctx is not (<fmpq_mpoly>other).ctx:
-                raise IncompatibleContextError(f"{(<fmpq_mpoly>self).ctx} is not {(<fmpq_mpoly>other).ctx}")
-            res = create_fmpq_mpoly(self.ctx)
-            fmpq_mpoly_sub(res.val, (<fmpq_mpoly>self).val, (<fmpq_mpoly>other).val, res.ctx.val)
-            return res
-        else:
-            other = any_as_fmpq(other)
-            if other is not NotImplemented:
-                res = create_fmpq_mpoly(self.ctx)
-                fmpq_mpoly_sub_fmpq(res.val, (<fmpq_mpoly>self).val, (<fmpq>other).val, res.ctx.val)
-                return res
-        return NotImplemented
-
-    def __rsub__(self, other):
-        cdef fmpq_mpoly res
-        other = any_as_fmpq(other)
-        if other is not NotImplemented:
-            res = create_fmpq_mpoly(self.ctx)
-            fmpq_mpoly_sub_fmpq(res.val, (<fmpq_mpoly>self).val, (<fmpq>other).val, res.ctx.val)
-            return -res
-        return NotImplemented
-
-    def __mul__(self, other):
-        cdef fmpq_mpoly res
-        if typecheck(other, fmpq_mpoly):
-            if (<fmpq_mpoly>self).ctx is not (<fmpq_mpoly>other).ctx:
-                raise IncompatibleContextError(f"{(<fmpq_mpoly>self).ctx} is not {(<fmpq_mpoly>other).ctx}")
-            res = create_fmpq_mpoly(self.ctx)
-            fmpq_mpoly_mul(res.val, (<fmpq_mpoly>self).val, (<fmpq_mpoly>other).val, res.ctx.val)
-            return res
-        else:
-            other = any_as_fmpq(other)
-            if other is not NotImplemented:
-                res = create_fmpq_mpoly(self.ctx)
-                fmpq_mpoly_scalar_mul_fmpq(res.val, (<fmpq_mpoly>self).val, (<fmpq>other).val, res.ctx.val)
-                return res
-        return NotImplemented
-
-    def __rmul__(self, other):
+    def _add_scalar_(self, other: fmpq):
         cdef fmpq_mpoly res
-        other = any_as_fmpq(other)
-        if other is not NotImplemented:
-            res = create_fmpq_mpoly(self.ctx)
-            fmpq_mpoly_scalar_mul_fmpq(res.val, (<fmpq_mpoly>self).val, (<fmpq>other).val, res.ctx.val)
-            return res
-        return NotImplemented
+        res = create_fmpq_mpoly(self.ctx)
+        fmpq_mpoly_add_fmpq(res.val, self.val, other.val, self.ctx.val)
+        return res
 
-    def __pow__(self, other, modulus):
+    def _add_mpoly_(self, other: fmpq_mpoly):
         cdef fmpq_mpoly res
```

Comment:
Maybe these methods should all be `cdef`.

Reply:
I've just tested this locally and it's possible, just needed a little massaging due to C-methods not being visible at runtime. The non-commutative operands can't use runtime reflection after coercion anymore, it needs a explicit C reflected function e.g. `_rtruediv_mpoly_` which can just call the normal C `_truediv_mpoly_` after a cast +  argument swap. Will have this up this evening

Reply:
Actually should just be able to use the normal dunder methods after the coercion rather than have a reflected cdef.

Reply:
[PR opened](https://github.com/flintlib/python-flint/pull/192)